sqlserver/datadog_checks/sqlserver/connection.py

```diff
@@ -0,0 +1,308 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+
+from contextlib import contextmanager
+
+from six import raise_from
+
+try:
+    import adodbapi
+except ImportError:
+    adodbapi = None
+
+try:
+    import pyodbc
+except ImportError:
+    pyodbc = None
+
+DATABASE_EXISTS_QUERY = 'select name from sys.databases;'
+
+
+class SQLConnectionError(Exception):
+    """Exception raised for SQL instance connection issues"""
+
+    pass
+
+
+class Connection(object):
+    """Manages the connection to a SQL Server instance."""
+
+    DEFAULT_COMMAND_TIMEOUT = 5
+    DEFAULT_DATABASE = 'master'
+    DEFAULT_DRIVER = 'SQL Server'
+    DEFAULT_DB_KEY = 'database'
+    PROC_GUARD_DB_KEY = 'proc_only_if_database'
+
+    valid_adoproviders = ['SQLOLEDB', 'MSOLEDBSQL', 'SQLNCLI11']
+    default_adoprovider = 'SQLOLEDB'
+
+    def __init__(self, init_config, check_instance):
+        self.instance = check_instance.instance
+        self.log = check_instance.log
```

Comment:
That is a good reference, but since this class needs to be initialized at check instance init, I think it's easier to just pass the logger directly.

Agree with your other comments about avoiding the need for the actual instance vs the config, though.